File: python/src/n0003_longest_substring_without_repeating_characters.py
from typing import Optional


def length_of_longest_substring(s: str) -> int:
    seen: list[Optional[int]] = [None] * 256
    max_length = 0
    start = 0
    for i, c in enumerate(s):
        prev = seen[ord(c)]
        if prev is not None:
            start = max(start, prev + 1)
        max_length = max(max_length, i - start + 1)
        seen[ord(c)] = i
    return max_length

File: rust/src/n0003_longest_substring_without_repeating_characters.rs
pub fn length_of_longest_substring(s: &str) -> usize {
    let mut max_len = 0;
    let mut start = 0;
    let mut seen = [None; 256];
    for (i, b) in s.bytes().enumerate() {
        if let Some(prev) = seen[b as usize] {
            start = start.max(prev + 1);
        }
        max_len = max_len.max(i - start + 1);
        seen[b as usize] = Some(i);
    }
    max_len
}
